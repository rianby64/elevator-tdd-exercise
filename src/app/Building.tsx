
import * as React from 'react';

import { Lift } from './Lift';
import { Floor } from './Floor';

interface BuildingProps {
    floors: number;
    DELAY?: number;
    onLiftAtFloor?: (i: number) => void;
}

interface DefaultProps {
    floors: number;
    DELAY: number;
}

interface BuildState {
    lift: {
        currentLevel: number;
    };
}

export class Building extends React.Component<BuildingProps & DefaultProps> {

    public static defaultProps: DefaultProps = {
        floors: 1,
        DELAY: 1000,
    }

    public state: BuildState = {
        lift: {
            currentLevel: 1,
        }
    }

    private moveLiftToFloor = (i: number): void => {
        (this.state.lift.currentLevel !== i) && setTimeout((): void => {
            if (this.state.lift.currentLevel !== i) {
                this.setState((state: BuildState): BuildState => {
                    let nextFloor = state.lift.currentLevel;
                    if (state.lift.currentLevel < i) {
                        nextFloor++;
                    }
                    if (state.lift.currentLevel > i) {
                        nextFloor--;
                    }
                    this.props.onLiftAtFloor && this.props.onLiftAtFloor(nextFloor);
                    return {...state, lift: { currentLevel: nextFloor }};
                });
                this.moveLiftToFloor(i);
            }
        }, this.props.DELAY);
    }

    private createFloors = (): JSX.Element[] => new Array(this.props.floors)
        .fill(undefined)
        .map((v: undefined, i: number, arr: undefined[]): JSX.Element => {
            if (i === 0) {
                return <Floor
                    currentLevel={this.state.lift.currentLevel}
                    key={i} level={i + 1}
                    firstLevel={true} />;
            }
            if (i + 1 === arr.length) {
                return <Floor
                    currentLevel={this.state.lift.currentLevel}
                    key={i} level={i + 1}
                    lastLevel={true} />;
            }
            return <Floor
                currentLevel={this.state.lift.currentLevel}
                key={i} level={i + 1}/>;
        })
        .reverse()

    private renderOneFloor(): JSX.Element {
        return (
            <div>This is just a House</div>
        );
    }

    public render(): JSX.Element {
        return this.props.floors == 1 ? this.renderOneFloor() : (
            <div>
                <span>Building</span>
                <Lift
                    onCallLift={this.moveLiftToFloor}
                    floors={this.props.floors}
                    currentLevel={this.state.lift.currentLevel} />
                {this.createFloors()}
            </div>
        );
    }
}
