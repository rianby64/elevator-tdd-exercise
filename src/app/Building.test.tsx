
import * as React from 'react';
import { mount } from 'enzyme';
import { Building } from './Building';

test('Building exists', (): void => {
    const el = (<Building />);
    const sh = mount(el);
    expect(sh).toMatchSnapshot();
});

test('Building draws two floors', (): void => {
    const el = (<Building floors={2} />);
    const sh = mount(el);
    expect(sh).toMatchSnapshot();
});

test('Building draws three floors', (): void => {
    const el = (<Building floors={3} />);
    const sh = mount(el);
    expect(sh).toMatchSnapshot();
});
