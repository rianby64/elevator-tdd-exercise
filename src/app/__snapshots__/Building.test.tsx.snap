// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Building draws three floors 1`] = `
<Building
  DELAY={1000}
  floors={3}
>
  <div>
    <span>
      Building
    </span>
    <Lift
      currentLevel={1}
      floors={3}
      onCallLift={[Function]}
    >
      <div>
        <LiftPanel
          callLift={[Function]}
          currentLevel={1}
          floors={3}
          highlightedButtons={Set {}}
        >
          <div>
            <span>
              Lift's panel
            </span>
            <ul>
              <li
                key="2"
              >
                <button
                  onClick={[Function]}
                >
                  3
                </button>
              </li>
              <li
                key="1"
              >
                <button
                  onClick={[Function]}
                >
                  2
                </button>
              </li>
              <li
                className="active"
                key="0"
              >
                <button
                  onClick={[Function]}
                >
                  1
                </button>
              </li>
            </ul>
          </div>
        </LiftPanel>
      </div>
    </Lift>
    <Floor
      currentLevel={1}
      firstLevel={false}
      key="2"
      lastLevel={true}
      level={3}
    >
      <div>
        <span>
          Floor 
          3
        </span>
        <Component
          downButton={true}
          upButton={false}
        >
          <ul>
            <li>
              Down
            </li>
          </ul>
        </Component>
      </div>
    </Floor>
    <Floor
      currentLevel={1}
      firstLevel={false}
      key="1"
      lastLevel={false}
      level={2}
    >
      <div>
        <span>
          Floor 
          2
        </span>
        <Component
          downButton={true}
          upButton={true}
        >
          <ul>
            <li>
              Up
            </li>
            <li>
              Down
            </li>
          </ul>
        </Component>
      </div>
    </Floor>
    <Floor
      currentLevel={1}
      firstLevel={true}
      key="0"
      lastLevel={false}
      level={1}
    >
      <div>
        <span
          className="active"
        >
          Floor 
          1
        </span>
        <Component
          downButton={false}
          upButton={true}
        >
          <ul>
            <li>
              Up
            </li>
          </ul>
        </Component>
      </div>
    </Floor>
  </div>
</Building>
`;

exports[`Building draws two floors 1`] = `
<Building
  DELAY={1000}
  floors={2}
>
  <div>
    <span>
      Building
    </span>
    <Lift
      currentLevel={1}
      floors={2}
      onCallLift={[Function]}
    >
      <div>
        <LiftPanel
          callLift={[Function]}
          currentLevel={1}
          floors={2}
          highlightedButtons={Set {}}
        >
          <div>
            <span>
              Lift's panel
            </span>
            <ul>
              <li
                key="1"
              >
                <button
                  onClick={[Function]}
                >
                  2
                </button>
              </li>
              <li
                className="active"
                key="0"
              >
                <button
                  onClick={[Function]}
                >
                  1
                </button>
              </li>
            </ul>
          </div>
        </LiftPanel>
      </div>
    </Lift>
    <Floor
      currentLevel={1}
      firstLevel={false}
      key="1"
      lastLevel={true}
      level={2}
    >
      <div>
        <span>
          Floor 
          2
        </span>
        <Component
          downButton={true}
          upButton={false}
        >
          <ul>
            <li>
              Down
            </li>
          </ul>
        </Component>
      </div>
    </Floor>
    <Floor
      currentLevel={1}
      firstLevel={true}
      key="0"
      lastLevel={false}
      level={1}
    >
      <div>
        <span
          className="active"
        >
          Floor 
          1
        </span>
        <Component
          downButton={false}
          upButton={true}
        >
          <ul>
            <li>
              Up
            </li>
          </ul>
        </Component>
      </div>
    </Floor>
  </div>
</Building>
`;

exports[`Building exists 1`] = `
<Building
  DELAY={1000}
  floors={1}
>
  <div>
    This is just a House
  </div>
</Building>
`;

exports[`Lift in Building goes from first to second floor 1`] = `
<Building
  DELAY={1}
  floors={3}
  onLiftAtFloor={[Function]}
>
  <div>
    <span>
      Building
    </span>
    <Lift
      currentLevel={1}
      floors={3}
      onCallLift={[Function]}
    >
      <div>
        <LiftPanel
          callLift={[Function]}
          currentLevel={1}
          floors={3}
          highlightedButtons={
            Set {
              2,
            }
          }
        >
          <div>
            <span>
              Lift's panel
            </span>
            <ul>
              <li
                key="2"
              >
                <button
                  onClick={[Function]}
                >
                  3
                </button>
              </li>
              <li
                key="1"
              >
                <button
                  className="highlight"
                  onClick={[Function]}
                >
                  2
                </button>
              </li>
              <li
                className="active"
                key="0"
              >
                <button
                  onClick={[Function]}
                >
                  1
                </button>
              </li>
            </ul>
          </div>
        </LiftPanel>
      </div>
    </Lift>
    <Floor
      currentLevel={1}
      firstLevel={false}
      key="2"
      lastLevel={true}
      level={3}
    >
      <div>
        <span>
          Floor 
          3
        </span>
        <Component
          downButton={true}
          upButton={false}
        >
          <ul>
            <li>
              Down
            </li>
          </ul>
        </Component>
      </div>
    </Floor>
    <Floor
      currentLevel={1}
      firstLevel={false}
      key="1"
      lastLevel={false}
      level={2}
    >
      <div>
        <span>
          Floor 
          2
        </span>
        <Component
          downButton={true}
          upButton={true}
        >
          <ul>
            <li>
              Up
            </li>
            <li>
              Down
            </li>
          </ul>
        </Component>
      </div>
    </Floor>
    <Floor
      currentLevel={1}
      firstLevel={true}
      key="0"
      lastLevel={false}
      level={1}
    >
      <div>
        <span
          className="active"
        >
          Floor 
          1
        </span>
        <Component
          downButton={false}
          upButton={true}
        >
          <ul>
            <li>
              Up
            </li>
          </ul>
        </Component>
      </div>
    </Floor>
  </div>
</Building>
`;

exports[`Lift in Building goes from first to third floor passing the second 1`] = `
<Building
  DELAY={1}
  floors={3}
  onLiftAtFloor={[Function]}
>
  <div>
    <span>
      Building
    </span>
    <Lift
      currentLevel={1}
      floors={3}
      onCallLift={[Function]}
    >
      <div>
        <LiftPanel
          callLift={[Function]}
          currentLevel={1}
          floors={3}
          highlightedButtons={
            Set {
              3,
            }
          }
        >
          <div>
            <span>
              Lift's panel
            </span>
            <ul>
              <li
                key="2"
              >
                <button
                  className="highlight"
                  onClick={[Function]}
                >
                  3
                </button>
              </li>
              <li
                key="1"
              >
                <button
                  onClick={[Function]}
                >
                  2
                </button>
              </li>
              <li
                className="active"
                key="0"
              >
                <button
                  onClick={[Function]}
                >
                  1
                </button>
              </li>
            </ul>
          </div>
        </LiftPanel>
      </div>
    </Lift>
    <Floor
      currentLevel={1}
      firstLevel={false}
      key="2"
      lastLevel={true}
      level={3}
    >
      <div>
        <span>
          Floor 
          3
        </span>
        <Component
          downButton={true}
          upButton={false}
        >
          <ul>
            <li>
              Down
            </li>
          </ul>
        </Component>
      </div>
    </Floor>
    <Floor
      currentLevel={1}
      firstLevel={false}
      key="1"
      lastLevel={false}
      level={2}
    >
      <div>
        <span>
          Floor 
          2
        </span>
        <Component
          downButton={true}
          upButton={true}
        >
          <ul>
            <li>
              Up
            </li>
            <li>
              Down
            </li>
          </ul>
        </Component>
      </div>
    </Floor>
    <Floor
      currentLevel={1}
      firstLevel={true}
      key="0"
      lastLevel={false}
      level={1}
    >
      <div>
        <span
          className="active"
        >
          Floor 
          1
        </span>
        <Component
          downButton={false}
          upButton={true}
        >
          <ul>
            <li>
              Up
            </li>
          </ul>
        </Component>
      </div>
    </Floor>
  </div>
</Building>
`;
